ksqldb-streams/src/test/java/io/confluent/ksql/execution/streams/ForeignKeyTableTableJoinBuilderTest.java

```diff
@@ -0,0 +1,315 @@
+      .keyColumn(R_KEY, SqlTypes.STRING)
+      .valueColumn(ColumnName.of("R_RED"), SqlTypes.BIGINT)
+      .valueColumn(ColumnName.of("R_ORANGE"), SqlTypes.DOUBLE)
+      .valueColumn(R_KEY, SqlTypes.STRING) // Copy of key in value
+      .build();
+
+  @Mock
+  private QueryContext ctx;
+  @Mock
+  private KTable<Struct, GenericRow> leftKTable;
+  @Mock
+  private KTable<Struct, GenericRow> leftKTableMultiKey;
+  @Mock
+  private KTable<Struct, GenericRow> rightKTable;
+  @Mock
+  private KTable<Struct, GenericRow> resultKTable;
+  @Mock
+  private ExecutionStep<KTableHolder<Struct>> left;
+  @Mock
+  private ExecutionStep<KTableHolder<Struct>> leftMultiKey;
+  @Mock
+  private ExecutionStep<KTableHolder<Struct>> right;
+  @Mock
+  private ExecutionKeyFactory<Struct> executionKeyFactory;
+  @Mock
+  private PlanInfo planInfo;
+  @Mock
+  private MaterializationInfo.Builder materializationBuilder;
+
+  private PlanBuilder planBuilder;
+  private ForeignKeyTableTableJoin<Struct, Struct> join;
+
+  @SuppressWarnings("unchecked")
+  @Before
+  public void init() {
+    when(left.build(any(), eq(planInfo))).thenReturn(
+        KTableHolder.materialized(leftKTable, LEFT_SCHEMA, executionKeyFactory, materializationBuilder));
+    when(leftMultiKey.build(any(), eq(planInfo))).thenReturn(
+        KTableHolder.materialized(leftKTableMultiKey, LEFT_SCHEMA_MULTI_KEY, executionKeyFactory, materializationBuilder));
+    when(right.build(any(), eq(planInfo))).thenReturn(
+        KTableHolder.materialized(rightKTable, RIGHT_SCHEMA, executionKeyFactory, materializationBuilder));
+
+    when(leftKTable.leftJoin(any(KTable.class), any(KsqlKeyExtractor.class), any())).thenReturn(resultKTable);
+    when(leftKTable.join(any(KTable.class), any(KsqlKeyExtractor.class), any())).thenReturn(resultKTable);
+    when(leftKTableMultiKey.leftJoin(any(KTable.class), any(KsqlKeyExtractor.class), any())).thenReturn(resultKTable);
+    when(leftKTableMultiKey.join(any(KTable.class), any(KsqlKeyExtractor.class), any())).thenReturn(resultKTable);
+
+    planBuilder = new KSPlanBuilder(
+        mock(RuntimeBuildContext.class),
+        mock(SqlPredicateFactory.class),
+        mock(AggregateParamsFactory.class),
+        mock(StreamsFactories.class)
+    );
+  }
+
+  @Test
+  public void shouldDoLeftJoinOnNonKey() {
+    // Given:
+    givenLeftJoin(left, JOIN_COLUMN);
+
+    // When:
+    final KTableHolder<Struct> result = join.build(planBuilder, planInfo);
+
+    // Then:
+    verify(leftKTable).leftJoin(
+        same(rightKTable),
+        eq(new KsqlKeyExtractor<>(1)),
+        eq(new KsqlValueJoiner(LEFT_SCHEMA.value().size(), RIGHT_SCHEMA.value().size(), 0))
+    );
+    verifyNoMoreInteractions(leftKTable, rightKTable, resultKTable);
+    assertThat(result.getTable(), is(resultKTable));
+    assertThat(result.getExecutionKeyFactory(), is(executionKeyFactory));
+  }
+
+  @Test
+  public void shouldDoLeftJoinOnKey() {
```

Comment:
What is the purpose of this test? Isn't this a primary key join, not a foreign key join?

Reply:
I guess you are right -- seems I overshot on corner case testing. Even if we could compile a PK_PK-join as a FK-join and it should still work -- it would be less efficient though I guess.

This actually raises a question: if we have two tables with different partition count and/or different key format, could we implement a PK-PK-join without repartitioning one input table (or even rejecting the join do the missing co-partitioning), but compiling it down as FK-join?

Thus, I am wondering if we should just keep this test, because from a physical plan point of view, this should work, even if we don't use it when translating the logical plan into a physical plan atm -- but maybe we would do this in the future?

Reply:
> if we have two tables with different partition count and/or different key format, could we implement a PK-PK-join without repartitioning one input table (or even rejecting the join do the missing co-partitioning), but compiling it down as FK-join?

I suppose we could but it seems odd to do this by default, as it allows users to inadvertently write queries with poorer performance than expected. I'm not convinced we need these unit tests until (and if) we decide to do this one day, but I don't feel strongly. If we're going to leave them in, can we at least add a comment explaining that these are actually primary key joins which won't reach this code today, but we've left it in the future? Otherwise future readers may be similarly confused.

Reply:
Agreed. Also not sure _if_ we should do it, but from a unit-test point of view, it seem ok to keep the tests, as unit testing the physical plan should not know anything about the logical plan layer.

Will add a comment to clarify.